File: ccc/ccc03s2.py
N = int(input())

def get_rhyme(word):
  vowels = {'a', 'e', 'i', 'o', 'u'}
  l = len(word)
  for i in range(l-1, -1, -1):  #starting from the end, go backwards
    if word[i] in vowels:
      break
  else:
    i = 0 #whole word
  return word[i:]

for _ in range(N):
  rhymes = []
  for _ in range(4):
    words = input().split()
    end = words[-1].lower()
    rhymes.append(get_rhyme(end))

  if rhymes[0] == rhymes[1] == rhymes[2] == rhymes[3]:
    print("perfect")
  elif rhymes[0] == rhymes[1] and rhymes[2] == rhymes[3]:
    print("even")
  elif rhymes[0] == rhymes[2] and rhymes[1] == rhymes[3]:
    print("cross")
  elif rhymes[0] == rhymes[3] and rhymes[1] == rhymes[2]:
    print("shell")
  else:
    print("free")
  

File: ccc/ccc03s2.cpp
#include <iostream>
#include <string>
using namespace std;

string get_rhyme(string words) {
    int size = words.size();
    int i;
    for (i = size - 1; i >= 0; i--) {
        if (words[i] == 'a' || words[i] == 'e' || words[i] == 'i' || words[i] == 'o' || words[i] == 'u' || words[i] == ' ')
            break;
    }
    if (i < 0) i = 0;
    return words.substr(i);
}

string to_lower(string s) {
    string t = "";
    for (auto c : s) {
        t += tolower(c);
    }
    return t;
}

int main() {
    int N;
    cin >> N;
    cin.ignore(1000, '\n');
    for (int i = 0; i < N; i++) {
        string rhymes[4];
        for (int j = 0; j < 4; j++) {
            string words;
            getline(cin, words);
            string rhyme = get_rhyme(to_lower(words));
            rhymes[j] = rhyme;
        }
        if (rhymes[0] == rhymes[1] && rhymes[1] == rhymes[2] && rhymes[2] == rhymes[3]) {
            cout << "perfect" << endl;
        }
        else if (rhymes[0] == rhymes[1] && rhymes[2] == rhymes[3]) {
            cout << "even" << endl;
        }
        else if (rhymes[0] == rhymes[2] && rhymes[1] == rhymes[3]) {
            cout << "cross" << endl;
        }
        else if (rhymes[0] == rhymes[3] && rhymes[1] == rhymes[2]) {
            cout << "shell" << endl;
        }
        else {
            cout << "free" << endl;
        }
    }
}